### ISSUE TITLE
Problem with the RateLimiter

### ISSUE THREAD

**Issue by xavivars** [sjdirect/abot#66]
Hi,

We've faced twice during last week a failure inside RateLimiter.This is the stack trace when the exception happened.
```
System.Management.Automation.RemoteException
Unhandled Exception: System.ArgumentOutOfRangeException: Number must be either non-negative and less than or equal to Int32.MaxValue or -1.
Parameter name: dueTime
   at System.Threading.Timer.Change(Int32 dueTime, Int32 period)
   at Abot.Util.RateLimiter.ExitTimerCallback(Object state)
   at System.Threading.ExecutionContext.runTryCode(Object userData)
   at System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, Object userData)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean ignoreSyncCtx)
   at System.Threading._TimerCallback.PerformTimerCallback(Object state)
```

We're using the RateLimiter to limit the requests per second, but this only happened after days of crawling...

**Comment by sjdirect** [comment]
Hi,

Did this happen while crawling a single very long site or multiple sites?
If later, do you create new instances of all the objects for each crawl or
do you reuse them? Is rate limiting happening due to a robots.txt file
giving a crawl-delay or are you setting the
minCrawlDelayPerDomainMilliSeconds config to a value greater than 0? Trying
to get an idea of the circumstances when this occurs.

Steven

**Comment by xavivars** [comment]
Hi Steven,
* This happened on a single very long site.
* The rate limiting occurs because we have our own implementation of IRateLimiter (called RangeRateLimiter) that does different "request per second" limitations depending on the time of the day. This object contains a list of Abot's RateLimiter, and uses one of them appropriately.

Below you can see an extract of the code 

```c#
class RangeRateLimiter: IRateLimiter
{
        public Dictionary<int, IRateLimiter> RateLimiters { get; set; }

        public void WaitToProceed()
        {
            IRateLimiter rateLimiter = GetRateLimiter(DateTime.Now);
            
            rateLimiter.WaitToProceed();
        }

         public IRateLimiter GetRateLimiter(DateTime dateTime)
        {
            int rate = GetRateForDateTime(dateTime);

            var secondTimeSpan = new TimeSpan(0, 0, 0, 1);
            if (!RateLimiters.ContainsKey(rate))
            {
                RateLimiters.Add(rate, new RateLimiter(rate, secondTimeSpan));
            }

            return RateLimiters[rate];
        }
}
```

Then, when we setup PoliteWebCrawler, we do something like 

```c#
IRateLimiter rangeRateLimiter = new RangeRateLimiter();
crawler.PageCrawlStarting += (s, e) => rangeRateLimiter.WaitToProceed();
```

But this happened after crawling more than 900k pages, after more than 5 days of crawling (and to give you an idea, we have 2 different "rates" right now: one during the day and one at night). So the object had already been used before.

**Comment by sjdirect** [comment]
Hi,

Not really enough information to go on yet. Please keep adding details to
this and i'll take a look when after we have more details.

**Comment by xavivars** [comment]
Sure. 

What other details you'd like to have?

**Comment by sjdirect** [comment]
If you can artificially duplicate the issue with a isolated block of code
that would be the best. I would try to create a loop that creates
RateLimiters and callis WaitToProceed or something like that. It will be a
while until I would have time to give this a deep dive.

**Comment by xavivars** [comment]
I've tried to replicate it with a dozen different crawls, with the same code + configuration, crawling similar (both in size, URL structure and infrastructure, and it only failed for two of them.

I'll try to do that

**Comment by sjdirect** [comment]
Haven't heard any new info to attempt to replicate. I will be closing this issue. However if you find a way to replicate it I'll reopen.

**Comment by xavivars** [comment]
Hi Steven,

I finally managed to reproduce (and isolate) the issue.

It seems to be related to this Danthar/RateLimiting/issues/1. Also, there are some comments on the original thread where the rate limiting code was proposed, and they suggest a fix for this.

http://www.jackleitch.net/2010/10/better-rate-limiting-with-dot-net/

There's also a comment from a Steven (maybe you?) talking about using a disposed rate limiter, with a locking proposal, and I've also found an alternative (easier) fix for the same issue here: http://stackoverflow.com/a/10526988/200472.

Does this make sense to you?
